README.md
# Cedi Circuitbreakers

Quick links:

- [About the library](#about)
- [Examples of use](#usage)
- [Configuration](#config)
- [How to get latest version](#getit)
- [API Docs](https://oss.sonatype.org/service/local/repositories/releases/archive/com/ccadllc/cedi/circuitbreaker_2.12/2.0.0/circuitbreaker_2.12-2.0.0-javadoc.jar/!/com/ccadllc/cedi/circuitbreaker/index.html)

### <a id="about"></a>About the library

The term `Circuit breaker` in software engineering applies to a programming pattern meant to provide a similar service for software components that its namesake provides for electrical grids: ensuring that a failure or overload does not cascade to bring down the whole system.  Advanced circuit breakers such as the ones provided by this library also enable the means for throttling (degrading or limiting) the component usage when a service cannot keep up with inbound requests.  Finally, this library additionally provides the means by which clients - for example, a monitoring task - can subscribe to 1.) events related to the circuit breaker activity (e.g., alert users when a circuit breaker has opened, closed, or is throttling requests due to excessive inbound traffic); and/or 2.) a stream of statistics related to the circuitbreaker emitted on a periodic basis.

### <a id="usage"></a>Examples of use

#### <a id="protectusage"></a> Circuitbreaker Protection

Here is a simple example of the use of a circuit breaker to protect against cascading failures. Aside from its configuration - described later in this README - the syntax and semantics in the use of the flow control circuit breaker is identical.

```scala
import cats.effect.IO
import fs2.Stream

import java.io.IOException
import java.util.UUID
import java.util.concurrent.TimeoutException

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.duration._

import com.ccadllc.cedi.circuitbreaker._

/*
 * The circuit breaker registry settings are fairly simple and only consist
 * of how often it should evaluate any of the circuit breakers and remove those
 * that have had no activity for a given time period.  If the checkInterval
 * property evaluates to 0 (0.seconds for example), no circuit breaker garbage
 * collection will occur.
 */
val circuitBreakerRegistrySettings: RegistrySettings = RegistrySettings(
  garbageCollection = RegistrySettings.GarbageCollection(checkInterval = 1.hour, inactivityCutoff = 24.hours)
)

/*
 * This constitutes the configuration of the circuitbreaker we shall create as part of this example.
 * It consists of the constraints that define sliding sample window of statistics to gather,
 * the percentage failures over that sample window time-frame which will trigger the circuitbreaker
 * to open and to fail fast subsequent requests, the interval at which a request should be let
 * through to test when the circuitbreaker has opened, and the number of consecutive test requests
 * which must succeed to close it.
 */
val databaseCircuitBreakerSettings: FailureSettings = FailureSettings(
  sampleWindow = SampleWindow(duration = 2.minutes),
  degradationThreshold = Percentage(percent = 40.0),
  test = FailureSettings.Test(interval = 5.seconds, minimumSuccesses = 5)
)

/* The very simple data type which will represent the database object we are querying for this example. */
case class QuarterlyProductSales(productId: UUID, productName: String, totalSales: Long)

/* A simple stub to represent a query to a remote dynamodb persistent database */
def retrieveQuarterlyProductSales: IO[Vector[QuarterlyProductSales]] =
  IO(Vector(QuarterlyProductSales(UUID.randomUUID, "widgets", 565000L)))

/*
 * A service to abstract interactions with a remote database.  For our example purposes, it only has one
 * entry point, to query for `QuarterlyProductSales` objects.
 */
class DatabaseService(cbRegistry: CircuitBreakerRegistry[IO], circuitBreakerSettings: FailureSettings) {
  /*
   * The dynamodb client circuit breaker is identified in the registry via the
   * `CircuitBreaker.Identifier`
   */
  final val DynamoDbCircuitBreakerId: CircuitBreaker.Identifier = CircuitBreaker.Identifier("dynamodb-client")
  /*
   * The dynamodb client circuit breaker will evaluate task failures using
   * this to determine if the failure should be considered in determining
   * whether a circuit breaker be opened. For example, application level failures
   * typically should not be considered, since that category of failure generally
   * does not indicate systemic problems of the kind a circuit breaker was designed
   * to protect against.
   */
  final val DynamoDbCircuitBreakerEvaluator: CircuitBreaker.FailureEvaluator = CircuitBreaker.FailureEvaluator {
    case _: IOException | _: TimeoutException => true
    case _ => false
  }

  def getAllQuarterlyProductSales: IO[Vector[QuarterlyProductSales]] = for {
    /*
     * The registry will look up the circuit breaker by identifier,
     * creating it if it does not yet exist.
     */
    cb <- cbRegistry.forFailure(DynamoDbCircuitBreakerId, circuitBreakerSettings, DynamoDbCircuitBreakerEvaluator)
    /*
     * The protect function of the circuit breaker will apply its current statistics
     * both before and after the task execution.  If the circuit breaker is open and
     * a test is not yet ready to be run, the task itself will not be executed but
     * will 'fail fast' with a `CircuitBreaker.OpenException`; Similarly, with a flow control
     * circuit breaker, if the rate per second is too high, requests will 'fail fast' with a
     * 'CircuitBreaker.ThrottledException' - and thus not be counted in the inbound rate stats -
     * until the inbound rate has been brought down to a level the processing rate statistics
     * indicates can be handled.
     */
    result <- cb.protect(retrieveQuarterlyProductSales)
  } yield result
}

/*
 * Near the beginning of the universe, create a circuit breaker registry for the
 * system to use. The circuit breaker is passed configuration settings as described
 * above.
 */
val protectedSystem = (for {
  cbRegistry <- Stream.eval(CircuitBreakerRegistry.create[IO](circuitBreakerRegistrySettings))
  dbService = new DatabaseService(cbRegistry, databaseCircuitBreakerSettings)
  result <- Stream.eval(dbService.getAllQuarterlyProductSales)
} yield ()).runLast.map(_.getOrElse(Vector.empty))

/*
 * Here would be the rest of the program in a real world application.
 */

/* At the end of the universe, run things! */
protectedSystem.unsafeRunSync()
```

#### <a id="eventsusage"></a> Circuit Breaker Event Subscription

Interested processes can subscribe to events related to a system's circuit breaker activity.  There is an example of its use.

```scala
import cats.effect.{Concurrent, IO}
import fs2.Stream
import com.ccadllc.cedi.circuitbreaker.{ CircuitBreakerRegistry, CircuitBreaker }
import com.ccadllc.cedi.circuitbreaker.statistics.{ FailureStatistics, FlowControlStatistics }

/**
 * In this method, we subscribe to a stream of circuit breaker events triggered on
 * state changes and process them in some manner (that manner is not shown here).
 */
def monitorCircuitBreakerEvents(cbRegistry: CircuitBreakerRegistry[IO]): IO[Unit] = {
  def processCircuitBreakerEvent(event: CircuitBreaker.CircuitBreakerEvent): IO[Unit] = event match {
    case CircuitBreaker.OpenedEvent(id: CircuitBreaker.Identifier, stats: statistics.FailureStatistics) => ???
    case CircuitBreaker.ClosedEvent(id: CircuitBreaker.Identifier, stats: statistics.FailureStatistics) => ???
    case CircuitBreaker.ThrottledDownEvent(id: CircuitBreaker.Identifier, stats: statistics.FlowControlStatistics) => ???
    case CircuitBreaker.ThrottledUpEvent(id: CircuitBreaker.Identifier, stats: statistics.FlowControlStatistics) => ???
  }
  val maxQueuedEvents = 100
  val eventStream = cbRegistry.events(maxQueuedEvents) flatMap { event =>
    Stream.eval_(processCircuitBreakerEvent(event))
  }
  Concurrent[IO].start(eventStream.run) map { _ => () }
}

/*
 * Near the beginning of the universe, create a circuit breaker registry for the
 * system to use. The circuit breaker is passed configuration settings, as described above.
 */
val protectedSystem = (for {
  cbRegistry <- Stream.eval(CircuitBreakerRegistry.create[IO](circuitBreakerRegistrySettings))
  dbService = new DatabaseService(cbRegistry, databaseCircuitBreakerSettings)
  result <- Stream.eval(dbService.getAllQuarterlyProductSales)
} yield ()).runLast.map(_.getOrElse(Vector.empty))

/*
 * Here would be the rest of the program in a real world application.
 */

/* At the end of the universe, run things! */
protectedSystem.unsafeRunSync()
```

### <a id="config"></a> Configuration

#### <a id="registryconfiguration"></a> Circuit Breaker Registry

The circuit breaker registry settings are defined in the `com.ccadllc.cedi.circuitbreaker` namespace and the defaults in this library's `reference.conf` generally do not need to be altered; however they are described below so that the library users understand their purpose.
```
com.ccadllc.cedi.circuitbreaker {
  registry {
    # The circuit breaker registry periodically checks to see if a
    # circuit breaker has been inactive for a lengthy period of time,
    # indicating it is no longer in use.  Since circuit breakers
    # use memory to maintain statistics, it is vital for long running
    # systems which create and discard circuit breakers that we
    # evict such inactive circuit breakers from the registry (they will
    # be re-created again if needed).  If a circuit breaker is inactive
    # for 60 minutes, by default, it shall be so evicted.
    # If check-interval evaluates to 0 (for instance, is specified as 0
    # seconds) then no garbage collection shall occur.
    garbage-collection {
      check-interval: 5 minutes
      inactivity-cutoff: 60 minutes
    }
  }
}
```

#### <a id="failureconfiguration"></a> Cascading Failure Circuit Breaker

The cascading failure circuit breaker has a number of configurable parameters, each of
which is described below.  The circuit breaker settings are typically defined inline subordinate
to the subsystem that uses them in the application configuration.  For example, for a DynamoDb
client circuit breaker in a `persistence` library, they would
be nested in the `dynamodb {  }` settings in the `persistence` reference.conf
file.

```
failure-circuitbreaker {
  # The sample window configuration determines the number of request
  # failure statistics that are maintained.  It is a sliding window.
  # For example, in this configuration, the system will maintain
  # request success/failure metrics for the last 2 minutes, with a
  # maximum of 50,000 entries.  The circuit breaker will only apply
  # protection once the sample window has initially been filled (the
  # first two minutes or 50,000 requests collected in this example).
  # Thereafter, the window continues to slide over the latest two
  # minutes of requests.
  # The maximum-entries entry is meant as a failsafe to ensure that
  # we hold no more than this number at any one time, to avoid
  # excessive memory consumption.
  sample-window {
    duration: 2 minutes
    maximum-entries: 50000
  }
  # This parameter indicates the percentage of failure detected by the
  # circuit breaker of the requests in the sample window which are
  # determine to be failures.  As described in the example below, a failure
  # of a protected task is determined by the `Evaluator` passed to the
  # given circuit breaker when created and may be unique to a given
  # subsystem.  For example, in this configuration, if 40% of the
  # requests processed in the latest 2 minute window have been deemed
  # failures, the circuit breaker will open and subsequent requests
  # will fail fast with a custom exception, with the exception of
  # those passed through for periodic testing.
  degradation-threshold {
    percent: 40.0
  }
  # Once a circuit breaker is open, we need a way to determine
  # that the failure condition(s) have been resolved so that
  # it can be closed again and normal request processing
  # resumed.  This is done by periodically letting a request
  # pass through and monitoring its result.  In this example,
  # when a circuit breaker has opened, a request will be passed
  # through every 30 seconds and if 5 requests in succession are
  # successful, the breaker will again open.
  test {
    interval: 30 seconds
    minimum-successes: 5
  }
  # This is set to true by default but if set to false, the circuit
  # breaker's failure protection will be disabled.  This defaults to
  # 'true' if not specified and is only specified here for illustration
  # purposes as it is not meant to be a 'public' parameter and should
  # generally be changed to 'false' only when troubleshooting.
  enabled: true
}
```

#### <a id="flowcontrolconfiguration"></a> Flow Control Circuit Breaker ###

The flow control circuit breaker has a number of configurable parameters, each of which is described below.  The circuit breaker settings are typically defined subordinate to the subsystem that uses them in the application configuration.  For example, for a circuit breaker protecting an http client in a `service` microservice, they would be nested in the `http-client {  }` settings in the `service` reference.conf file.  See the [typesafe documentation](https://github.com/typesafehub/config) for more details on the general usage of this type of configuration.

```
flow-control-circuitbreaker {
  # Since the throttling circuit breaker also has cascading failure protection, it includes those
  # configuration parameters, as in the previous section above.  Note that if the 'enabled'
  # parameter is set to 'false' in the failure sub-section of the flow control circuit breaker
  # configuration, it will disable failure protection; however, flow control protection will
  # still be enabled unless the flow control 'enabled' parameter is set to 'false.'
  failure {
    sample-window {
      duration: 2 minutes
      maximum-entries: 50000
    }
    degradation-threshold {
      percent: 40.0
    }
    test {
      interval: 30 seconds
      minimum-successes: 5
    }
    enabled: true
  }
  # This sliding sample window maintains the mean rate per second of
  # both inbound requests and requests processed, so it can compare the two
  # to determine if the protected task is keeping up by ensuring the requests
  # processed per second does not fall behind the inbound rate in a sustained manner.
  # For that reason, the sample window here is usually larger than the failure
  # sample window described above as you typically want a larger sample size by
  # which to factor out bursts of traffic versus sustained rates. The circuit
  # breaker protection will not be applied until the sample window has initially
  # been filled (the first 30 seconds in this example). Thereafter, the window
  # continues to slide over the latest 30 seconds of inbound and processing rates.
  # A maximum-entries value can also be specified here; however, in general it
  # is not needed since the maximum entries will never be more than the number of
  # seconds for the sample window (in this case, 30).  It is only if you have a
  # very, very long window that you would need to specify it to override the
  # calculation based on the sample window.
  sample-window {
    duration: 30 seconds
  }

  # In order to ensure we avoid throttling requests unnecessarily, and to further
  # guard against bursts of traffic, you can configure a circuit breaker such that
  # it will not start throttling requests until the observed mean processing
  # rate per second has fallen behind the inbound rate by the percentage configured
  # here.  If set to zero, the mean processing rate must equal the mean inbound
  # rate or throttling will begin.  Because of bursting and timing, You typically
  # want this value to be non-zero to avoid overly aggressive throttling. In the
  # example below, the processing rate must fall behind the inbound rate by 10
  # percent.  Only performance testing and tuning can determine the appropriate
  # value for a given system.

  allowed-over-processing-rate {
    percent: 10.0
  }

  # This value is used to determine whether or not to trigger a throttled up or
  # throttled down event, in order to avoid a flood of events while
  # throttling is in effect.  So in this configuration, an event would only been
  # triggered when throttling up 5% (allowing 5% more requests per second through)
  # or throttling down 5% (throttling - failing fast - 5% more requests per
  # second).
  minimum-reportable-rate-change {
    percent: 5.0
  }
  # This is set to true by default but if set to false, the circuit
  # breaker's flow control protection will be disabled.  This defaults to
  # 'true' if not specified and is only specified here for illustration
  # purposes as it is not meant to be a 'public' parameter and should
  # generally be changed only when troubleshooting.  Note that if this is
  # set to 'false' but the failure sub-config has its enabled set to 'true'
  # failure protection will still be enabled and the circuit breaker will
  # in effect act like a standard failure protection only variant.
  enabled: true

  # This indicates the hard limit for inbound requests per second.  If at any time the
  # number of requests per second exceed this value, requests will get throttled
  # until the start of the next second.  This is meant as a "failsafe" to
  # prevent sudden spikes from overwhelming a protected service prior to the
  # adaptive measures kicking in (the input versus mean processing rates in
  # the sample window).  In general, it should be set higher than the protected
  # service can be handled in a sustained manner but not so high that it will
  # overwhelm the system for the period of time that it takes for the sample window
  # mean values to adjust to the increase in load and provide adaptive protection.
  per-second-rate-threshold: 25
}
```

### <a id="getit"></a>How to get latest Version

Cedi Circuitbreaker supports 2.11, 2.12, and 2.13. It is published on Maven Central.

```scala
libraryDependencies += "com.ccadllc.cedi" %% "circuitbreaker" % "2.0.0"
```

## Copyright and License

This project is made available under the [Apache License, Version 2.0](LICENSE). Copyright information can be found in [NOTICE](NOTICE).


CHANGELOG.md
2.0.0
=====
 - Added support for scala 2.13 and library refresh, including fs2 and cats 2.x


core/src/test/scala/com/ccadllc/cedi/circuitbreaker/FailureCircuitBreakerTest.scala
/*
 * Copyright 2016 Combined Conditional Access Development, LLC.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.ccadllc.cedi.circuitbreaker

import cats.effect.IO

import org.scalatest.wordspec.AnyWordSpec

import scala.concurrent.duration._

class FailureCircuitBreakerTest extends AnyWordSpec with TestSupport {
  "The failure circuit breaker" should {
    "switch to the open position when the failure rate of the services it is protecting exceeds the configured threshold" in {
      val id = CircuitBreaker.Identifier("test")
      val registry = CircuitBreakerRegistry.create[IO](testRegistryConfig).unsafeRunSync
      val failureThreshold = Percentage(20.0)
      val cb = registry.forFailure(id, testFailureConfig.copy(degradationThreshold = failureThreshold)).unsafeRunSync
      val tseo = TestStreamedEventObserver.create(registry)
      tseo.openedCount(id) shouldBe 0
      val results = protectFailure(cb, failureThreshold.plus(Percentage(10.0)))
      tseo.openedCount(id) should be > 0
      results.collectFirst { case Some(CircuitBreaker.OpenException(eId, _)) => eId } shouldBe Some(id)
    }
    "switch to the closed position after opening when the configured minimum number of service tests succeed" in {
      val id = CircuitBreaker.Identifier("test")
      val registry = CircuitBreakerRegistry.create[IO](testRegistryConfig).unsafeRunSync
      val failureThreshold = Percentage(20.0)
      val cb = registry.forFailure(
        id,
        testFailureConfig.copy(degradationThreshold = failureThreshold, test = testFailureConfig.test.copy(interval = 0.seconds))).unsafeRunSync
      val tseo = TestStreamedEventObserver.create(registry)
      tseo.closedCount(id) shouldBe 0
      protectFailure(cb, failureThreshold.plus(Percentage(10.0)))
      tseo.openedCount(id) should be > 0
      protectFailure(cb, Percentage.minimum)
      tseo.closedCount(id) should be > 0
    }
  }
}


core/src/main/scala/com/ccadllc/cedi/circuitbreaker/package.scala
/*
 * Copyright 2016 Combined Conditional Access Development, LLC.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.ccadllc.cedi

/**
 * The circuitbreaker library provides protection against cascading failure and
 * system overload via the [[CircuitBreaker]] and its primary function [[CircuitBreaker#protect]].
 * A [[CircuitBreakerRegistry]] maintains the collection of active circuit breakers and is
 * the interface through which circuit breakers are created, retrieved, and removed and by
 * which clients can subscribe for state change and statistics events related to the
 * circuit breakers.
 */
package object circuitbreaker {
  /**
   * Provides syntax enrichment on the effectful program `F[A]` so that
   * one can write, for example, `prg.protect(circuitBreaker)` rather than
   * `circuitBreaker.protect(prg)`.  This can lead to less awkward code
   * under certain circumstances.
   */
  implicit class CircuitBreakerOps[F[_], A](val self: F[A]) extends AnyVal {
    /**
     * Alternate manner in which to protect the effectful program `F[A]`.
     * See [[CircuitBreaker#protect]] for details.
     *
     * @param cb - the [[CircuitBreaker]] instance which will protect this `F[A]`.
     * @return enhancedProgram - the protected effectful program.
     */
    def protect(cb: CircuitBreaker[F]): F[A] = cb.protect(self)
  }
}


core/src/main/scala/com/ccadllc/cedi/circuitbreaker/StateRef.scala
/*
 * Copyright 2016 Combined Conditional Access Development, LLC.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.ccadllc.cedi.circuitbreaker

import cats.effect.Sync
import cats.implicits._

import java.util.concurrent.atomic.AtomicReference

/*
 * Provides lock-free atomic retrieval and update capabilities for an unconstrained immutable
 * data type.  Used by this library to maintain state in the [[CircuitBreaker]] instances and the [[CircuitBreakerRegistry]].
 */
private[circuitbreaker] final class StateRef[F[_], A] private (ref: AtomicReference[A])(implicit F: Sync[F]) {
  /* Retrieve the state */
  def get: F[A] = F.delay(ref.get)
  /* Modify the state with the passed-in function, returning the modified value */
  def modify(f: A => A): F[A] = F.delay(ref.updateAndGet(a => f(a)))
  /*
   * Retrieve a sub-component `B` of the state `A` with the `retriever` function,
   * creating the component `B` with the `creator` function if it does not exist in `A`
   * (modifying `A` with the newly created `B` using the `toA` function), returning an
   * effectful program that when run will produce the desired component `B`.
   */
  def getOrCreate[B](retriever: A => Option[B], creator: F[B], toA: (A, B) => A): F[B] = {
    def createB: F[B] = for {
      cb <- creator
      a <- modify(a => retriever(a).fold(toA(a, cb))(_ => a))
      b <- retriever(a).fold(createB)(F.pure)
    } yield b
    for {
      a <- get
      bMaybe <- F.pure(retriever(a))
      b <- bMaybe.fold(createB)(F.pure)
    } yield b
  }
}
private[circuitbreaker] object StateRef {
  def create[F[_], A](value: => A)(implicit F: Sync[F]): F[StateRef[F, A]] = F.delay(new StateRef[F, A](new AtomicReference(value)))
}


core/src/main/scala/com/ccadllc/cedi/circuitbreaker/CircuitBreakerSettings.scala
/*
 * Copyright 2016 Combined Conditional Access Development, LLC.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.ccadllc.cedi.circuitbreaker

import scala.concurrent.duration._

import com.ccadllc.cedi.config._
import ConfigParser._

/**
 * Represents a percentage value, used throughout the library.
 */
case class Percentage(percent: Double) {
  import Percentage._
  require(percent >= Minimum && percent <= Maximum, s"Percent must be between $Minimum and $Maximum inclusive but was $percent")

  /**
   * Add two percentages together, ensuring that the result will be within the limits
   * of the [[Percentage#Minimum]] and [[Percentage#Maximum]].
   * @param other - the percentage to add to `this`
   * @return combinedPercentage - the sum total of the two percentages,
   *   within the limits of [[Percentage#Minimum]] and [[Percentage#Maximum]].
   */
  def plus(other: Percentage): Percentage = Percentage.withinLimits(percent + other.percent)

  /**
   * Subtract two percentages, ensuring that the result will be within the limits
   * of the [[Percentage#Minimum]] and [[Percentage#Maximum]].
   * @param other - the percentage to subtract from `this`
   * @return subtractedPercentage - the result of subtracting the `other` percentage from `this`,
   *   within the limits of [[Percentage#Minimum]] and [[Percentage#Maximum]].
   */
  def minus(other: Percentage): Percentage = Percentage.withinLimits(percent - other.percent)

  /**
   * Is the passed-in percentage greater than `this`?
   * @param other - the percentage to compare
   * @return isGreater - the passed-in percentage is (or isn't) greater than `this`.
   */
  def greaterThan(other: Percentage): Boolean = percent > other.percent

  /**
   * The inverse of `this`.  That is, subtract 100.00 from the value of `this`.
   * @return inversePercentage - the result of subtracting 100.00 from the value of `this`.
   */
  def inverse: Percentage = Percentage.withinLimits(100.0 - percent)

  /**
   * human-readable rendering of the `Percentage`.
   */
  def show: String = s"$percent%"
}

object Percentage {

  /** The minimum value of a `Percentage` */
  final val Minimum: Double = 0.0

  /** The maximum value of a `Percentage` */
  final val Maximum: Double = 100.0

  /** Used to parse from `HCON` formatted configuration files */
  implicit val configParser: ConfigParser[Percentage] = derived[Percentage]

  /** commonly used value representing a minimum percentage */
  val minimum: Percentage = Percentage(Minimum)

  /** Is the value within our imposed percentage limits? */
  def withinLimits(value: Double): Percentage = Percentage(value.min(Maximum).max(Minimum))
}

/**
 * Represents a window of time, defined by the `duration` property and
 * used to constrain the number of time-stamped sample items retained in order
 * to calculate statistics for deriving [[CircuitBreaker]] state changes.  A
 * hard limit of the number of samples is defined by the `maximumEntries`
 * property.
 */
case class SampleWindow(duration: FiniteDuration, maximumEntries: Int = Int.MaxValue) {
  require(maximumEntries > 0, s"maximum entries must be a positive number")
  def show: String = s"${duration.toCoarsest} with failsafe maximum entries of $maximumEntries"
}
object SampleWindow {
  implicit val configParser: ConfigParser[SampleWindow] = (
    duration("duration") ~ int("maximum-entries").optional) map { case (dur, mes) => SampleWindow(dur, mes.getOrElse(dur.toSeconds.toInt + 1)) }
}

/**
 * The configuration used for the [[CircuitBreakerRegistry]] - it defines how often the
 * registry should examine the [[CircuitBreaker]]s under its control to remove instances
 * which have been inactive longer than the configured cut-off time.  If the `checkInterval`
 * property of the [[RegistrySettings#GarbageCollection]] configuration item is set to a
 * 0 value (e.g., 0.milliseconds), this feature will then be disabled.
 */
case class RegistrySettings(garbageCollection: RegistrySettings.GarbageCollection) {
  def show: String = s"garbage collection [ ${garbageCollection.show} ]"
}
object RegistrySettings {
  case class GarbageCollection(checkInterval: FiniteDuration, inactivityCutoff: FiniteDuration) {
    def show: String = s"check interval [ ${checkInterval.toCoarsest} ], inactivity cutoff [ ${inactivityCutoff.toCoarsest} ]"
  }
  object GarbageCollection {
    implicit val configParser: ConfigParser[GarbageCollection] = derived[GarbageCollection]
  }
  implicit val configParser: ConfigParser[RegistrySettings] =
    subconfig("com.ccadllc.cedi.circuitbreaker.registry")(derived[RegistrySettings])
}

/**
 * The configuration which controls the failure-based [[CircuitBreaker]]s (that is, those circuit breakers
 * which are created via the [[CircuitBreakerRegistry#forFailure]] function).
 * @param sampleWindow - the time-based window to be used to constrain the number of program execution
 *   success/failure indicators to store in order to determine the percentage of failures over this
 *   time period.  An example would be `SampleWindow(2.minutes)` which means maintain the last two minutes
 *   of program execution success/failure indicators for a program protected by a given [[CircuitBreaker]].
 * @param degradationThreshold - the [[Percentage]] of items within the sample window which indicate program
 *   failure.  For instance, if the percentage were `Percentage(40.0)` and there were 50 sample items in the
 *   2 minute window, with 25 indicating program failure, then since 50% is greater than 40%, the circuit breaker
 *   would open.
 * @param test - the testing configuration is used when the circuit breaker is open in order to determine how
 *   often a program should be executed rather than failed fast and then how many consecutively successful
 *   test executions need to occur in order to close the circuit breaker and allow normal operation of the
 *   protected program.
 * @param enabled - The [[CircuitBreaker]] can be enabled or disabled with this parameter.
 */
case class FailureSettings(
  sampleWindow: SampleWindow,
  degradationThreshold: Percentage,
  test: FailureSettings.Test,
  enabled: Boolean = true) {
  def show: String = s"sample window [ ${sampleWindow.show} ], degradation threshold [ ${degradationThreshold.show} ], test [ ${test.show} ], enabled [ $enabled ]"
}
object FailureSettings {
  case class Test(interval: FiniteDuration, minimumSuccesses: Int) {
    def show: String = s"interval [ ${interval.toCoarsest} ], minimum successes [ $minimumSuccesses ]"
  }
  object Test {
    implicit val configParser: ConfigParser[Test] = (duration("interval") ~ int("minimum-successes")) map { case (i, ms) => Test(i, ms) }
  }
  implicit val configParser: ConfigParser[FailureSettings] = (
    subconfig("sample-window")(SampleWindow.configParser) ~
    subconfig("degradation-threshold")(Percentage.configParser) ~
    subconfig("test")(Test.configParser) ~
    bool("enabled").withFallbackDefault(true)) map { case (sw, dt, t, e) => FailureSettings(sw, dt, t, e) }
}

/**
 * The configuration which controls the flow control-based [[CircuitBreaker]]s (that is, those circuit breakers
 * which are created via the [[CircuitBreakerRegistry#forFlowControl]] function).
 * @param failure - the configuration for the failure component of the flow control [[CircuitBreaker]].  In addition
 *   to providing flow control, it also provides the same failure functionality as its failure-only variant.
 * @param sampleWindow - the time-based window to be used to constrain the number of inbound program execution
 *   requests per second and program executions processed per second.  The flow control circuit breaker maintains
 *   two separate collections (the inbound rate and the processing/completed rate) and both use this sample window
 *   configuration to constrain the sizes of these collections.
 * @param allowedOverProcessingRate - the [[Percentage]] that the mean inbound rate is allowed to exceed the mean processing
 *   rate before this circuit breaker starts throttling inbound requests to lower it to an acceptable rate.  Put another
 *   way, the acceptable rate is the processing rate + (processing rate * percentage / 100).
 * @param minimumReportableRateChange - this is the percent that the calculated acceptable rate must be observed to have
 *   changed (either lower or higher) before a state change event is triggered.  If it is 0, all state changes result
 *   in an event.  This configuration item is meant to alleviate a flood of events when the request traffic pattern
 *   is very choppy (lots of spikes and valleys in a short time period).
 * @param perSecondRateMinimum - this is the minimum observed inbound rate at which flow control throttling will kick in.
 * @param perSecondRateHardLimit - this is the hard limit of the inbound requests per second above which requests will be
 *   throttled by the circuit breaker (failed fast without execution) for the remainder of the current second.  It will
 *   be monitored and acted upon continuously.
 * @param enabled - The flow control portion of the [[CircuitBreaker]] can be enabled or disabled with this parameter.
 */
case class FlowControlSettings(
  failure: FailureSettings,
  sampleWindow: SampleWindow,
  allowedOverProcessingRate: Percentage,
  minimumReportableRateChange: Percentage,
  perSecondRateMinimum: Long,
  perSecondRateHardLimit: Long,
  enabled: Boolean = true) {
  def show: String = s"failure [ {failure.show} ], sample window [ ${sampleWindow.show} ], allowed over processing rate [ ${allowedOverProcessingRate.show} ], minimum reportable rate change [ ${minimumReportableRateChange.show} ], per second rate minimum [ $perSecondRateMinimum ], per second rate hard limit [ $perSecondRateHardLimit ], enabled [ $enabled ]"
}
object FlowControlSettings {
  implicit val configParser: ConfigParser[FlowControlSettings] = (
    subconfig("failure")(FailureSettings.configParser) ~
    subconfig("sample-window")(SampleWindow.configParser) ~
    subconfig("allowed-over-processing-rate")(Percentage.configParser) ~
    subconfig("minimum-reportable-rate-change")(Percentage.configParser) ~
    long("per-second-rate-minimum").withFallbackDefault(1L) ~
    long("per-second-rate-hard-limit") ~
    bool("enabled").withFallbackDefault(true)) map { case (f, sw, aopr, mrrc, psrm, psrhl, e) => FlowControlSettings(f, sw, aopr, mrrc, psrm, psrhl, e) }
}
